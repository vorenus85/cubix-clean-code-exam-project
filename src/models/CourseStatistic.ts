export class CourseStatistic{
    constructor(
        private courseName:string,
        private totalLectures:number,
        private lecturesCompleted:number,
        private progress:number,
        private lastAccessed:Date
        ) { }
    }