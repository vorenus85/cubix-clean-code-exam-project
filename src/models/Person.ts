import { PERSON_ERROR_MSG } from '../utils/personErrorMessages'

export abstract class Person {
  constructor(
    private name: string,
    private email: string
  ) {
    this.SetName(name)
    this.SetEmail(email)
  }

  public GetName(): string {
    return this.name
  }

  public GetEmail(): string {
    return this.email
  }

  private SetName(value: string) {
    if (!value || typeof value !== 'string') {
      throw new Error(PERSON_ERROR_MSG.NAME_EMPTY)
    } else {
      this.name = value
    }
  }

  private SetEmail(email: string) {
    if (!this.IsValidEmail(email)) {
      throw new Error(PERSON_ERROR_MSG.INVALID_EMAIL)
    } else {
      this.email = email
    }
  }

  private IsValidEmail(email: string): boolean {
    // Basic email validation using a regular expression
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
    return emailRegex.test(email)
  }
}
