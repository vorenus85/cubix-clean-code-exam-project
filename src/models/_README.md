The "models" folder is where classes or modules representing data structures, entities, or domain objects are stored. These classes often define the structure of the data used within the application. In the context of databases, these models might map to tables or documents.
