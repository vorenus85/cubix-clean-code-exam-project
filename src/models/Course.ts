import{ Student} from"./Student";

export class Course{ private students:Student[] =[];
    constructor( 
        private courseName: string, 
        private startDate: Date, 
        private lengthInWeeks: number, 
        private costInHuf: number) { }
    }