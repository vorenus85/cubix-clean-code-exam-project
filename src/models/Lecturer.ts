import { Course } from './Course'
import { Person } from './Person'

export class Lecturer extends Person {
  constructor(
    name: string,
    email: string,
    public assignedCourses: string[] = []
  ) {
    super(name, email)
  }
}
