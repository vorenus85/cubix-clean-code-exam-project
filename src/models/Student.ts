import { Course } from './Course'
import { Person } from './Person'

export class Student extends Person {
  constructor(
    name: string,
    email: string,
    public registeredCourses: string[] = []
  ) {
    super(name, email)
  }
}
