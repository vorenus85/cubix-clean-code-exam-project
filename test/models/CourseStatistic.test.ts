import { CourseStatistic } from '../../src/models/CourseStatistic'
import { CourseStatisticFixture } from '../fixtures'

describe('Course Statistic tests', () => {
  let courseStatistic: CourseStatistic
  const courseName = 'Clean Code'
  const totalLectures = 6
  const lecturesCompleted = 4
  const progress = 80
  const lastAccessed = new Date('2024-01-22')

  describe('Happy paths', () => {
    it('should create CourseStatistic with valid ', () => {
      courseStatistic = new CourseStatistic(
        courseName,
        totalLectures,
        lecturesCompleted,
        progress,
        lastAccessed
      )

      expect(courseStatistic.GetCourseName()).toBe(courseName)
      expect(courseStatistic.GetTotalLecture()).toBe(totalLectures)
      expect(courseStatistic.GetLecturesCompleted()).toBe(lecturesCompleted)
      expect(courseStatistic.GetProgress()).toBe(progress)
      expect(courseStatistic.GetLastAccessed()).toBe(lastAccessed)
    })

    test('should CourseStatistic exits with all possible properties', () => {
      const actualResult = CourseStatisticFixture

      expect(actualResult).toMatchSnapshot()
    })
  })
  describe('Error paths', () => {})
})
