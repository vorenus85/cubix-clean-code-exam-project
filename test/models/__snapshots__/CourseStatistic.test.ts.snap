// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Statistic tests Happy paths should Course exits with all possible properties 1`] = `
CourseStatistic {
  "courseName": "Clean Code",
  "lastAccessed": 2024-01-22T00:00:00.000Z,
  "lecturesCompleted": 4,
  "progress": 80,
  "totalLectures": 6,
}
`;

exports[`Course Statistic tests Happy paths should CourseStatistic exits with all possible properties 1`] = `
CourseStatistic {
  "courseName": "Clean Code",
  "lastAccessed": 2024-01-22T00:00:00.000Z,
  "lecturesCompleted": 4,
  "progress": 80,
  "totalLectures": 6,
}
`;
